### ISSUE TITLE
Log numbering problems

### ISSUE THREAD

**Issue by eurobank** [CIDRAM/CIDRAM#128]
Checking last month logs i see the following:

ID: 7288 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:58:56 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxxxxxxxxxx »

ID: 1 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:58:56 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: https://xxxxxxxxxxxxxxxx »

I do not understand why we started from ID 1 ?

Any ideas ?

**Comment by eurobank** [comment]
More strange numbering, double same IDs, reset to 1 and then continue to the previous .....

ID: 7278 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:54:32 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxx »

ID: 7278 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:54:32 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxxxxxxxxxxxx »

ID: 1 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:54:32 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxxxxxxxxxxxxxx »

ID: 7279 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:54:32 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxxxxxxxxxxxxxx »

ID: 7279 »
Script Version: CIDRAM v1.12.0 »
Date/Time: Wed, 22 May 2019 08:54:32 +0300 »
IP Address: 13.66.184.79 »
Signatures Count: 1 »
Signatures Reference: 13.64.0.0/14 »
Why Blocked: Cloud service ("Azure", L24803:F0, [US])! »
Reconstructed URI: xxxxxxxxxxxxxxxxxxxxx »

**Comment by htjoe38** [comment]
I've occasionally seen similar behavior when there is multiple hits per second - as your second post indicated. All happening at 08:54:32 If you were to check the logs you would probably see more hits happening at the same second. How many hits per second before this starts to show up is dependent on your server, and the hiccup in the log numbering I think happens as the server is trying to catch up with demand.

I'm sure others here will have a better/alternative explanation.

**Comment by eurobank** [comment]
@htjoe38 

Could be. Since i use one central CIDRAM installation for all my hosted sites in the hosting plan (6). 

Meaning multiple connections from multiple sites at the same time. 

Could also be that i prepend Ninja standalone firewall after the Cidram ?. Who knows ....

**Comment by dev-101** [comment]
I do not closely follow CIDRAM code development (due to lack of time), but there can be a considerable difference between server's and PHP's timestamp values. Also, PHP version might not take microsecond precision into account, so that can complicate and make things even worse for later tracing.

I solved this in one of my setups by using server's timestamp instead. Maybe that idea can be applied here, too.

**Comment by 737simpilot** [comment]
When I've gotten multiple hits in one second the log number is just the same. Like 3321 and 3321, and another 3321 log number all in one second. Which would be normal if I think so. In your case perhaps you had like a layer 7 DDoS type thing going on. CloudFlare can prevent this. And CIDRAM has this feature as well, but CloudFlare would probably be better.

**Comment by Maikuolan** [comment]
> In your case perhaps you had like a layer 7 DDoS type thing going on. CloudFlare can prevent this. And CIDRAM has this feature as well, but CloudFlare would probably be better.

Regarding DDoS attacks: I doubt that's what's happened here, but in terms of the effectiveness of CIDRAM versus CloudFlare versus whatever else, there is a part in our FAQ, which could help elaborate on that point a little: *"[Will CIDRAM protect my website from DDoS attacks?](https://github.com/CIDRAM/Docs/blob/master/readme.en.md#DDOS_ATTACKS)"*. Worth reading for anyone that hasn't already done so.

**Comment by Maikuolan** [comment]
But yeah.. most likely caused due to race conditions between multiple concurrent PHP processes, all executing CIDRAM at the same time, all instances trying to read/write the same files at the same time, effectively trying to lock each other out and causing the files to corrupt and/or reset entirely.

Not much way to avoid the problem as long as we continue using flatfiles for caching and logging. The supplementary cache options were intended to resolve that problem, by introducing some alternative caching mechanisms such as Redis, APCu, using PDO to store data, etc. The documentation stills advises against using the supplementary cache options at the moment though, until I can investigate and test everything a bit more in live environments, to be able to document some of the possible "gotchas" associated with these alternative caching mechanisms. I've noticed that the behaviour of some of these mechanisms doesn't seem to be entirely consistent between different hosting environments, and although generally able to resolve most of the race conditions associated with flatfiles, I've seen some other new problems introduced sometimes that didn't exist before. You could try experimenting with these options, if you feel up to it; It *might* resolve the problem for your particular environment, but *also* might similarly introduce other new, but similar problems with caching and logging (so, I would advice caution and closely monitoring, if you choose to experiment with these options).

From a technical perspective, the problem with flatfile race conditions is simple to observe and easy understand, but difficult to resolve without often much more complex solutions, I think.

**Comment by dev-101** [comment]
I see now the code at these lines:

https://github.com/CIDRAM/CIDRAM/blob/v2/vault/outgen.php#L353

```
if ($CIDRAM['BlockInfo']['SignatureCount'] > 0) {
        ...
        /** Fetch current cached log event counter. */
        $CIDRAM['BlockInfo']['Counter'] = $CIDRAM['Cache']->getEntry('Counter') ?: 0;
        ...
}
```

Later, there is this section:

```
/** If any signatures were triggered, log it, generate output, then die. */
if ($CIDRAM['BlockInfo']['SignatureCount'] > 0) {
    ...
    /** Build fields. */
    if (!empty($CIDRAM['BlockInfo']['Counter']) || $CIDRAM['Config']['general']['empty_fields'] === 'include') {
        if (empty($CIDRAM['BlockInfo']['Counter'])) {
            $CIDRAM['BlockInfo']['Counter'] = 0;
        }
        $CIDRAM['AddField']($CIDRAM['L10N']->getString('field_id'), $CIDRAM['BlockInfo']['Counter']);
    }
    ...
}
```

I wonder if $CIDRAM['Cache']->getEntry('Counter') is critical here and must be fetched from Cache (e.g. refers to particular signature/module ID or something) or it can be generated fresh? You could store counter value in some external file, and later use to read it, but I'm not sure about efficiency of that method in case of high traffic / many requests.

**Comment by eurobank** [comment]
Today i did this: i put my IP in the custom blocked IPs and then i executed a browsing script i created, simultaneously to all  my sites, some 32000 times.

CIDRAM worked great, logged everything, numbering is ok. My plan was "limited" from CloudLinux at some point, but my hoster was aware what i was doing.

So it is not a problem with multiple logging or a hosting problem or a "race" condition, IMHO.

**Comment by DanielRuf** [comment]
Imo we should not use these numbers but timestamps for the number. It is not uncommon that there are race conditions (can be also seen in some logfiles like accesslog).

But using something different than a normal counter which is standardized would reduce issues (well, we should still use monolog instead of reinventing the wheel regarding logging, log rotation and so on).

**Comment by Maikuolan** [comment]
> Imo we should not use these numbers but timestamps for the number. It is not uncommon that there are race conditions (can be also seen in some logfiles like accesslog).

When it *works* properly, and as intended, it provides a convenient way to refer to a specific log entry (e.g., when debugging a false positive, or when a user asks why something is being blocked, and that user then shares an entire log file, or a large chunk of a log file containing multiple entries, when we look at that data, we can say to them, *"please refer to log entry X, Y, Z..." and so on*), which is the main benefit of having log entry IDs.

The problem, however.. is when it *doesn't* work properly (it's kind of useless, to say *"please refer to log entry X, Y, Z..."*, when there are multiple instances of X, Y, Z in the same log file, due to race conditions or ID counters that regularly reset).

That said, however.. it's merely a convenience; it isn't (or shouldn't be) essential for debugging (as long as whoever does the debugging, knows how to read the log files properly), and we can (or should be able to) live without it, when it can't be relied on because of problems like these.

Edit: Forgot to mention. Log entries also already include timestamps (or.. the actual date/time, to be more specific; as opposed to a Unix-style timestamp), so, if we were to use timestamps instead, it would mostly be a matter of changing our approach to reading log data, rather than necessarily needing to change the way that data is logged (i.e., looking at the timestamps when debugging, instead of looking at the entry IDs).

**Comment by DanielRuf** [comment]
Normal integers could be problematic in some cases when there is an overflow in our code.

At least we have native `hrtime` in PHP 7.3+ now.

**Comment by DanielRuf** [comment]
All the solutions that I use (monolog, NinjaFirewall and so on) use some sort of random value + timestamps as there is a lower risk to have a collision than heaving race conditions which happen way more often - especially when you stream events and one event takes longer than another because of I/O and so on.

**Comment by DanielRuf** [comment]
> Date/Time: Wed, 22 May 2019 08:58:56 +0300

This is not a real timestamp in a normal context. Seconds as precision is often not enough. `microtime()` and others for getting the Unix timestamps are normally used in a logging context.

**Comment by DanielRuf** [comment]
> Date/Time: Wed, 22 May 2019 08:58:56 +0300

This is not a real timestamp in a logging context. Seconds as precision is often not enough. `microtime()` and others for getting the Unix timestamps are normally used in a logging context.

Let's make this more precise + random fraction to prevent such cases.

**Comment by Maikuolan** [comment]
This issue effectively resolved now by #131, I think (i.e., not relying on a counter anymore = no problems with counters being reset anymore). So, we should be okay to close this now. The cache resetting (e.g., #120) and files being deleted for no apparent reason (e.g., #127) are still an ongoing issues that need to be resolved somehow, but at least, we can address those particular issues at their respectively issue pages accordingly.
